addons/mail/static/src/models/throttle.js

```diff
@@ -0,0 +1,273 @@
+        },
+        /**
+         * @private
+         * @returns {string}
+         */
+        _computeId() {
+            return _.uniqueId('throttle_');
+        },
+        /**
+         * Called when there is a call to the function. This function is throttled,
+         * so the time it is called depends on whether the "cooldown stage" occurs
+         * or not:
+         *
+         * - no cooldown stage: function is called immediately, and it starts
+         *      the cooldown stage when successful.
+         * - in cooldown stage: function is called when the cooldown stage has
+         *      ended from timeout.
+         *
+         * Note that after the cooldown stage, only the last attempted function
+         * call will be considered.
+         *
+         * @private
+         * @param {...any} args
+         * @throws {ThrottleReinvokedError|ThrottleCanceledError}
+         * @returns {any} result of called function, if it's called.
+         */
+        async _do(...args) {
+            if (!this.isCoolingDown) {
+                return this._invokeFunction(...args);
+            }
+            if (this.pendingInvokeDeferred) {
+                this.pendingInvokeDeferred.reject(new ThrottleReinvokedError(this.id));
+            }
+            try {
+                this.update({ pendingInvokeDeferred: makeDeferred() });
+                await Promise.race([this.coolingDownDeferred, this.pendingInvokeDeferred]);
+            } catch (error) {
+                if (
+                    !(error instanceof ThrottleFlushedError) ||
+                    error.throttleId !== this.id
+                ) {
+                    throw error;
+                }
+            } finally {
+                this.update({ pendingInvokeDeferred: clear() });
+            }
+            return this._invokeFunction(...args);
+        },
+        /**
+         * Invoke the inner function of this throttle and starts cooling down phase
+         * immediately after.
+         *
+         * @private
+         * @param  {...any} args
+         */
+        async _invokeFunction(...args) {
+            const res = this.func(...args);
+            this._startCoolingDown();
+            return res;
+        },
+        /**
+         * Called just when the inner function is being called. Starts the cooling
+         * down phase, which turn any call to this throttle function as pending
+         * inner function calls. This will be called after the end of cooling down
+         * phase (except if canceled).
+         */
+        async _startCoolingDown() {
+            if (this.coolingDownDeferred) {
+                throw new Error("Cannot start cooling down if there's already a cooling down in progress.");
+            }
+            // Keep local reference of cooling down deferred, because the one stored
+            // on `this` could be overwritten by another call to this throttle.
+            const coolingDownDeferred = makeDeferred();
```

Comment:
This would probably be more clear as a field, and timer function as a non-arrow function (necessary to use timer model anyway)